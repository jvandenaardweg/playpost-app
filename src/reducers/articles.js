export const GET_ARTICLES = 'articles/LOAD';
export const GET_ARTICLES_SUCCESS = 'articles/LOAD_SUCCESS';
export const GET_ARTICLES_FAIL = 'articles/LOAD_FAIL';

export function articlesReducer(state = { articles: [] }, action) {
  switch (action.type) {
    case GET_ARTICLES:
      return { ...state, isLoading: true };
    case GET_ARTICLES_SUCCESS:
      return { ...state, isLoading: false, articles: action.payload.data };
    case GET_ARTICLES_FAIL:
      return {
        ...state,
        isLoading: false,
        error: 'Error while fetching the articles'
      };
    default:
      return state;
  }
}

export function listArticles(user) {
  return {
    type: GET_ARTICLES,
    payload: {
      request: {
        url: `/users/${user}/repos`
      }
    }
  };
}
