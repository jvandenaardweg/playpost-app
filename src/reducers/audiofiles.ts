
export const SET_DOWNLOADED_AUDIOFILE = 'audiofiles/SET_DOWNLOADED_AUDIOFILE';
export const RESET_STATE = 'audiofiles/RESET_STATE';

export type AudiofilesState = Readonly<{
  isLoading: boolean;
  downloaded: ReadonlyArray<Api.Audiofile>;
  error: string;
}>;

export const initialState: AudiofilesState = {
  isLoading: false,
  downloaded: [],
  error: ''
};

/* tslint:disable-next-line no-any */
export function audiofilesReducer(state = initialState, action: any): AudiofilesState {
  switch (action.type) {
    case SET_DOWNLOADED_AUDIOFILE:
      const audiofile: Api.Audiofile = action.payload;

      return {
        ...state,
        isLoading: false,
        downloaded: [
          ...state.downloaded.slice(0, 0),
          audiofile,
          ...state.downloaded.slice(0)
        ],
        error: ''
      };

    case RESET_STATE:
      return {
        ...initialState
      };

    default:
      return state;
  }
}

export const resetAudiofilesState = () => ({
  type: RESET_STATE
});

export const setDownloadedAudiofile = (audiofile: Api.Audiofile) => ({
  type: SET_DOWNLOADED_AUDIOFILE,
  payload: audiofile
});
