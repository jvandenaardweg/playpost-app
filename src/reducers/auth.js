import Analytics from 'appcenter-analytics';

export const POST_AUTH = 'auth/POST_AUTH';
export const POST_AUTH_SUCCESS = 'auth/POST_AUTH_SUCCESS';
export const POST_AUTH_FAIL = 'auth/POST_AUTH_FAIL';
export const REMOVE_AUTH = 'auth/REMOVE_AUTH';
export const SET_AUTH_TOKEN = 'auth/SET_AUTH_TOKEN';

const POST_AUTH_FAIL_MESSAGE = 'An unknown error happened while loggin you in. Please contact us when this happens all the time.';

const initialState = {
  isLoading: false,
  token: null,
  error: null
};

export function authReducer(state = initialState, action) {
  switch (action.type) {
    case POST_AUTH:
      return {
        ...state,
        isLoading: true
      };

    case POST_AUTH_SUCCESS:
      Analytics.trackEvent('Auth success');

      return {
        ...state,
        isLoading: false,
        token: action.payload.data.token,
        error: null
      };

    case POST_AUTH_FAIL:
      if (action.error.response && action.error.response.data && action.error.response.data.message) {
        Analytics.trackEvent('Error auth', { message: action.error.response.data.message });
      } else {
        Analytics.trackEvent('Error auth', { message: POST_AUTH_FAIL_MESSAGE });
      }

      return {
        ...state,
        isLoading: false,
        token: null,
        error: (action.error.response) ? action.error.response.data.message : POST_AUTH_FAIL_MESSAGE
      };

    case REMOVE_AUTH:
      return {
        ...initialState
      };

    case SET_AUTH_TOKEN:
      return {
        ...state,
        token: action.payload.token
      };

    default:
      return state;
  }
}

export function setAuthToken(token) {
  return {
    type: SET_AUTH_TOKEN,
    payload: {
      token
    }
  };
}

export function removeAuth() {
  return {
    type: REMOVE_AUTH
  };
}

export function postAuth(email, password) {
  return {
    type: POST_AUTH,
    payload: {
      request: {
        method: 'post',
        url: '/v1/auth',
        data: {
          email,
          password
        }
      }
    }
  };
}
