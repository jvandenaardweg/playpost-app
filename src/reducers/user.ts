import DeviceInfo from 'react-native-device-info';

import {
  CREATE_USER_FAIL_MESSAGE,
  DELETE_USER_FAIL_MESSAGE,
  GENERIC_NETWORK_ERROR,
  GET_USER_FAIL_MESSAGE,
  PATCH_USER_FAIL_MESSAGE,
  SAVE_SELECTED_VOICE_FAIL_MESSAGE
} from '../constants/messages';

export const GET_USER = 'user/GET_USER';
export const GET_USER_SUCCESS = 'user/GET_USER_SUCCESS';
export const GET_USER_FAIL = 'user/GET_USER_FAIL';

export const PATCH_USER = 'user/PATCH_USER';
export const PATCH_USER_SUCCESS = 'user/PATCH_USER_SUCCESS';
export const PATCH_USER_FAIL = 'user/PATCH_USER_FAIL';

export const CREATE_USER = 'user/CREATE_USER';
export const CREATE_USER_SUCCESS = 'user/CREATE_USER_SUCCESS';
export const CREATE_USER_FAIL = 'user/CREATE_USER_FAIL';

export const DELETE_USER = 'user/DELETE_USER';
export const DELETE_USER_SUCCESS = 'user/DELETE_USER_SUCCESS';
export const DELETE_USER_FAIL = 'user/DELETE_USER_FAIL';

export const SAVE_SELECTED_VOICE = 'user/SAVE_SELECTED_VOICE';
export const SAVE_SELECTED_VOICE_SUCCESS = 'user/SAVE_SELECTED_VOICE_SUCCESS';
export const SAVE_SELECTED_VOICE_FAIL = 'user/SAVE_SELECTED_VOICE_FAIL';
export const RESET_SAVE_SELECTED_VOICE_ERROR = 'user/RESET_SAVE_SELECTED_VOICE_ERROR';


export const SET_PLAYBACK_SPEED = 'user/SET_PLAYBACK_SPEED';

export const RESET_USER_STATE = 'user/RESET_USER_STATE';
export const RESET_USER_ERROR = 'user/RESET_USER_ERROR';

export type UserState = Readonly<{
  isLoading: boolean;
  isLoadingDelete: boolean;
  isLoadingPatchUser: boolean;
  isLoadingSaveSelectedVoice: boolean;
  details: Api.User | null;
  error: string;
  errorSaveSelectedVoice: string;
  deviceLocale: string;
  playbackSpeed: number;
}>;

export const initialState: UserState = {
  isLoading: false,
  isLoadingDelete: false,
  isLoadingPatchUser: false,
  isLoadingSaveSelectedVoice: false,
  details: null,
  error: '',
  errorSaveSelectedVoice: '',
  deviceLocale: DeviceInfo.getDeviceLocale(),
  playbackSpeed: 1
};

/* tslint:disable no-any */
export function userReducer(state = initialState, action: any): UserState {
  switch (action.type) {
    case GET_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case GET_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case GET_USER_FAIL:
      let getUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        getUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          getUserFailMessage = action.error.response.data.message;
        } else {
          getUserFailMessage = GET_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        error: getUserFailMessage
      };

    case CREATE_USER:
      return {
        ...state,
        isLoading: true,
        error: ''
      };

    case CREATE_USER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        details: action.payload.data,
        error: ''
      };

    case CREATE_USER_FAIL:
      let createUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        createUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          createUserFailMessage = action.error.response.data.message;
        } else {
          createUserFailMessage = CREATE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoading: false,
        details: null,
        error: createUserFailMessage
      };

    case DELETE_USER:
      return {
        ...state,
        isLoadingDelete: true,
        error: ''
      };

    case DELETE_USER_SUCCESS:
      return {
        ...state,
        isLoadingDelete: false,
        error: ''
      };

    case DELETE_USER_FAIL:
      let deleteUserFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        deleteUserFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          deleteUserFailMessage = action.error.response.data.message;
        } else {
          deleteUserFailMessage = DELETE_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingDelete: false,
        error: deleteUserFailMessage
      };

    case PATCH_USER:
      return {
        ...state,
        isLoadingPatchUser: true,
        error: ''
      };

    case PATCH_USER_SUCCESS:
      return {
        ...state,
        isLoadingPatchUser: false,
        error: ''
      };

    case PATCH_USER_FAIL:
      let updateUserEmailFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        updateUserEmailFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          updateUserEmailFailMessage = action.error.response.data.message;
        } else {
          updateUserEmailFailMessage = PATCH_USER_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingPatchUser: false,
        error: updateUserEmailFailMessage
      };

    case SAVE_SELECTED_VOICE:
      return {
        ...state,
        isLoadingSaveSelectedVoice: true,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_SUCCESS:
      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: ''
      };

    case SAVE_SELECTED_VOICE_FAIL:
      let saveSelectedVoiceFailMessage = '';

      // Network error
      if (action.error && action.error.status === 0) {
        saveSelectedVoiceFailMessage = GENERIC_NETWORK_ERROR;
      } else {
        // Error, from the API
        if (action.error && action.error.response && action.error.response.data && action.error.response.data.message) {
          saveSelectedVoiceFailMessage = action.error.response.data.message;
        } else {
          saveSelectedVoiceFailMessage = SAVE_SELECTED_VOICE_FAIL_MESSAGE;
        }
      }

      return {
        ...state,
        isLoadingSaveSelectedVoice: false,
        errorSaveSelectedVoice: saveSelectedVoiceFailMessage
      };

    case SET_PLAYBACK_SPEED:
      return {
        ...state,
        playbackSpeed: action.payload
      };

    case RESET_SAVE_SELECTED_VOICE_ERROR:
      return {
        ...state,
        errorSaveSelectedVoice: ''
      };

    case RESET_USER_ERROR:
      return {
        ...state,
        error: ''
      };

    case RESET_USER_STATE:
      return {
        ...initialState
      };

    default:
      return state;
  }
}

export function resetUserState() {
  return {
    type: RESET_USER_STATE
  };
}

export function resetSaveSelectedVoiceError() {
  return {
    type: RESET_SAVE_SELECTED_VOICE_ERROR
  };
}

export function resetUserError() {
  return {
    type: RESET_USER_ERROR
  };
}

export function getUser() {
  return {
    type: GET_USER,
    payload: {
      request: {
        method: 'get',
        url: '/v1/me'
      }
    }
  };
}

export function patchUser({ email, password }: { email?: string, password?: string }) {
  return {
    type: PATCH_USER,
    payload: {
      request: {
        method: 'patch',
        url: '/v1/me',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function createUser(email: string, password: string) {
  return {
    type: CREATE_USER,
    payload: {
      request: {
        method: 'post',
        url: '/v1/users',
        data: {
          email,
          password
        }
      }
    }
  };
}

export function deleteUser() {
  return {
    type: DELETE_USER,
    payload: {
      request: {
        method: 'delete',
        url: '/v1/me'
      }
    }
  };
}

/**
 * Saves the selected voice as a default voice for the voice's language
 *
 * @param voiceId
 */
export function saveSelectedVoice(voiceId: string) {
  return {
    type: SAVE_SELECTED_VOICE,
    payload: {
      request: {
        method: 'post',
        url: '/v1/me/voices',
        data: {
          voiceId
        }
      }
    }
  };
}

export function setPlaybackSpeed(speed: number) {
  return {
    type: SET_PLAYBACK_SPEED,
    payload: speed
  };
}
