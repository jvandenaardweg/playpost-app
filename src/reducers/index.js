import { articlesReducer } from './articles';
import { playerReducer } from './player';
import { authReducer } from './auth';
import { usersReducer } from './users';

const rootReducer = {
  articles: articlesReducer,
  player: playerReducer,
  auth: authReducer,
  users: usersReducer
};

export default rootReducer;
