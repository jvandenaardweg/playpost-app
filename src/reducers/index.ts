import { combineReducers, Action } from 'redux';
import { AxiosResponse, AxiosRequestConfig } from 'axios';

import { playerReducer } from './player';
import { authReducer } from './auth';
import { userReducer } from './user';
import { playlistReducer } from './playlist';
import { audiofilesReducer } from './audiofiles';
import { voicesReducer } from './voices';

/* tslint:disable no-any */
export interface AxiosActionResponse extends AxiosResponse {
  payload: {
    data: any
  };
}

export interface AxiosAction extends Action {
  payload: {
    request: AxiosRequestConfig
  };
}

const rootReducer = combineReducers({
  player: playerReducer,
  auth: authReducer,
  user: userReducer,
  playlist: playlistReducer,
  audiofiles: audiofilesReducer,
  voices: voicesReducer
});

export type RootState = ReturnType<typeof rootReducer>;

export { rootReducer };
