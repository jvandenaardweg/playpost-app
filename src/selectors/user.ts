import { Platform } from 'react-native';
import { createSelector } from 'reselect';
import { SUBSCRIPTION_NAME, SUBSCRIPTION_PRODUCT_ID_FREE } from '../constants/in-app-purchase';
import { RootState } from '../reducers';
import { UserState } from '../reducers/user';
import { createDeepEqualSelector } from './index';

export interface UserSelectedVoiceByLanguageName {
  [key: string]: Api.Voice
}

export const userSelector = (state: RootState): UserState => state.user;

export const selectUserError = createSelector(
  [userSelector],
  user => user.error
);

export const selectUserErrorSaveSelectedVoice = createSelector(
  [userSelector],
  user => user.errorSaveSelectedVoice
);

export const selectUserIsLoading = createSelector(
  [userSelector],
  user => user.isLoading
);

export const selectUserDetails = createSelector(
  [userSelector],
  user => user.details
);

export const selectUserUsedInAppSubscriptionTrials = createSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.usedInAppSubscriptionTrials
);

export const selectDeviceLocale = createSelector(
  [userSelector],
  user => {
    if (!user.deviceLocale) {
      return '';
    }

    // device locale could be: "en-NL" or "en"
    // So we always split the dash
    // Even if the dash does not exist, we return the first in the array
    const deviceLocaleSplitted = user.deviceLocale.split('-');

    if (!deviceLocaleSplitted || !deviceLocaleSplitted[0]) {
      return '';
    }

    return deviceLocaleSplitted[0];
  }
);

export const selectUserSelectedVoices = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails) => {
    if (!userDetails || !userDetails.voiceSettings || !userDetails.voiceSettings.length) {
      return [];
    }
    return userDetails.voiceSettings.map(userVoiceSetting => userVoiceSetting.voice);
  }
);

export const selectUserSelectedVoiceByLanguageName = createDeepEqualSelector(
  [selectUserSelectedVoices],
  (voices): UserSelectedVoiceByLanguageName | null => {
    // Convert the array to an object
    // So we can easily pick a language inside our components
    const voicesObjectWithLanguageNameKeys: UserSelectedVoiceByLanguageName = voices.reduce((prev, curr) => {
      prev[curr.language.name] = curr;

      return prev;
    }, {})

    if (!Object.keys(voicesObjectWithLanguageNameKeys).length) {
      return null;
    }

    return voicesObjectWithLanguageNameKeys;
  }
);

export const selectUserSubscriptions = createDeepEqualSelector(
  [selectUserDetails],
  userDetails => userDetails && userDetails.inAppSubscriptions
);

export const selectUserIsSubscribed = createDeepEqualSelector(
  [selectUserDetails],
  (userDetails): boolean => !!(userDetails && userDetails.activeUserInAppSubscription)
);

/**
 * The active subscription's productId is different per platform.
 *
 */
export const selectUserActiveSubscriptionProductId = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): string => {
    if (!isSubscribed) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    if (!userDetails.activeUserInAppSubscription.inAppSubscription) { return SUBSCRIPTION_PRODUCT_ID_FREE; }
    return userDetails.activeUserInAppSubscription.inAppSubscription.productId
  }
);

export const selectActiveUserInAppSubscription = createDeepEqualSelector(
  [selectUserDetails, selectUserIsSubscribed],
  (userDetails, isSubscribed): Api.UserInAppSubscriptionApple | Api.UserInAppSubscriptionGoogle | null => {
    if (!isSubscribed) { return null; }
    if (!userDetails) { return null; }
    if (!userDetails.activeUserInAppSubscription) { return null; }

    return userDetails.activeUserInAppSubscription;
  }
);

export const selectUserActiveSubscriptionName = createDeepEqualSelector(
  [selectActiveUserInAppSubscription],
  (userActiveSubscription): string => {
    if (!userActiveSubscription) {
      return SUBSCRIPTION_NAME[SUBSCRIPTION_PRODUCT_ID_FREE];
    }
    return userActiveSubscription.inAppSubscription.name
  }
);

export const selectUserIsEligibleForTrial = createDeepEqualSelector(
  [selectUserUsedInAppSubscriptionTrials],
  (usedInAppSubscriptionTrials): boolean => {
    if (!usedInAppSubscriptionTrials || !usedInAppSubscriptionTrials.length) {
      return true;
    }

    // If the platform is Android, look up if the user had a Google subscription trial before
    if (Platform.OS === 'android') {
      return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'google')
    }

    // If the platform is else (iOS), look up if the user had a Apple subscription trial before
    return !!!usedInAppSubscriptionTrials.find(usedInAppSubscription => usedInAppSubscription.service === 'apple')
  }
);

export const selectUserHasSubscribedBefore = createDeepEqualSelector(
  [selectUserSubscriptions],
  userInAppSubscriptions => {
    if (!userInAppSubscriptions || !userInAppSubscriptions.length) {
      return false;
    }

    // If the user has a previous in app subscription record, return true
    return !!userInAppSubscriptions.length;
  }
);

export const selectUserPlaybackSpeed = createSelector(
  [userSelector],
  user =>  user.playbackSpeed
)
