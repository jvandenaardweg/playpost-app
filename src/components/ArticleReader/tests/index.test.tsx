import React from 'react';
import { render, RenderAPI } from 'react-native-testing-library';
import { ArticleReader } from '../index';

import articleMock from '../../../../tests/__mocks__/article';
import articleMockRTL from '../../../../tests/__mocks__/article-right-to-left';
import { UserTheme } from '../../../reducers/user';

const defaultProps = {
  article: articleMock
}

describe('ArticleReader', () => {
  let wrapper: RenderAPI;

  describe('rendering default', () => {
    const props = {
      ...defaultProps
    }

    beforeAll(() => {
      wrapper = render(<ArticleReader {...props}></ArticleReader>);
    });

    it('should render correctly', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render text from left to right as default', () => {
      const textAlignOccurrences: string[] = wrapper.getByName('WebView').props.source.html.match(/direction: ltr;/g);

      expect(textAlignOccurrences.length).toBeDefined();

      textAlignOccurrences.forEach((textAlign) => {
        expect(textAlign.trim()).toBe('direction: ltr;')
      })
    });
  });

  describe('rendering right to left content', () => {
    const props = {
      ...defaultProps,
      article: articleMockRTL
    }

    beforeAll(() => {
      wrapper = render(<ArticleReader {...props}></ArticleReader>);
    });

    it('should render correctly', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render text from right to left', () => {
      const textAlignOccurrences: string[] = wrapper.getByName('WebView').props.source.html.match(/direction: rtl;/g);

      expect(textAlignOccurrences.length).toBeDefined();

      textAlignOccurrences.forEach((textAlign) => {
        expect(textAlign.trim()).toBe('direction: rtl;')
      })
    });
  });

  describe('rendering dark theme', () => {

    beforeAll(() => {
      wrapper = render(<ArticleReader article={articleMock} forceTheme={UserTheme.dark}></ArticleReader>);
    });

    it('should render correctly', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('rendering no html', () => {

    beforeAll(() => {
      articleMock.html = '';
      wrapper = render(<ArticleReader article={articleMock}></ArticleReader>);
    });

    it('should render a different HTML output when the article has not HTML', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('rendering no author name', () => {

    beforeAll(() => {
      articleMock.authorName = '';
      wrapper = render(<ArticleReader article={articleMock}></ArticleReader>);
    });

    it('should render no authorName when it is not present', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('rendering no article URL', () => {

    beforeAll(() => {
      articleMock.url = '';
      articleMock.canonicalUrl = '';
      wrapper = render(<ArticleReader article={articleMock}></ArticleReader>);
    });

    it('should render no article URL link', () => {
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });
});
