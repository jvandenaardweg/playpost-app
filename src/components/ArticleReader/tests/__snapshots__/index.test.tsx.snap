// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleReader rendering should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "overflow": "hidden",
    }
  }
>
  <RNCWKWebView
    bounces={true}
    cacheEnabled={true}
    decelerationRate={0.998}
    javaScriptEnabled={false}
    messagingEnabled={false}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onNavigationStateChange={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      Object {
        "baseUrl": "",
        "html": "
      <!DOCTYPE html>
      
      <head>
        <meta http-equiv=\\"content-type\\" content=\\"text/html; charset=utf-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width\\">
        <link href=\\"https://fonts.googleapis.com/css?family=PT+Sans:400,400i,700,700i|PT+Serif:400,400i,700,700i\\" rel=\\"stylesheet\\">
        <style type=\\"text/css\\">
          html {
            box-sizing: border-box;
          }
          *, *:before, *:after {
            box-sizing: inherit;
          }
          body {
            padding: 8px;
            font-size: 17px;
            font-family: 'PT Serif', serif;
            line-height: 1.5;
            color: #000000;
            font-weight: normal;
            word-break: break-word;
            background-color: #ffffff;
          }

          h1, h2, h3, h4, h5, h6, h7, h8 {
            line-height: 1.2;
            margin-top: 0;
            margin-bottom: 10px;
            color: #000000;
            font-family: 'PT Sans', sans-serif;
          }

          h1 {
            font-size: 24px;
            margin-bottom: 16px;
          }

          h2, h3, h4 {
            margin-top: 24px;
            font-size: 21px;
          }

          h5, h6, h7, h8 {
            margin-top: 24px;
            font-size 18px
          }

          p {
            margin-top: 0;
            text-align: justify;
          }

          a, strong {
            color: #000000;
          }

          img, figure {
            max-width: 100%;
            margin-bottom: 16px;
            margin-top: 16px;
            display: block;
          }

          blockquote {
            font-style: italic;
            color: #000000;
            margin-left: 24px;
            margin-right: 24px;
          }
          blockquote p {
            font-style: italic;
          }

          figcaption {
            display: none;
          }

          .meta-header {
            margin-bottom: 24px;
          }

          .meta-header strong {
            display: block;
            color: #888888;
            font-weight: normal;
          }

        </style>
      </head>
        <body>
          <h1>Strategies to Pull Yourself Out of a Bad Mood</h1>
          <div class=\\"meta-header\\">
            <strong>Cindy Lamothe</strong>
            <strong><a href=\\"https://medium.com/s/story/strategies-to-pull-yourself-out-of-a-bad-mood-db31d5709fff\\">medium.com</a></strong>
          </div>
          <div id=\\"readability-page-1\\" class=\\"page\\"><div><div><div><div><p><span>T</span>he incident that ruined my day happened early in the morning: I was in my car and inadvertently cut off another driver, who then zoomed up next to me and yelled, “Idiot!” outside my window before hastily driving off. Frustrated, I yelled something nasty back, even though I knew he couldn’t hear me — which frustrated me, too.</p><p>Even minor annoyances like these can create a domino effect. The first annoyance magnifies each one that follows, all of them building on each other until you’ve worked yourself into a full-on bad mood.</p><p>“When you’re in a bad mood or having a bad day, it’s easy to chalk it up to bad luck,” says Kate Hanley, a personal development coach and author of <a href=\\"https://katehanley.com/books/\\"><em>How to Be a Better Person</em></a>, “which makes it feel as if you’re helpless to do anything about it.”</p><p>But there are ways you can successfully lessen your negative reactions before they escalate. When you feel the day spiraling away from you, simply noticing that you’re in a funk is an important first step toward getting out of it. Here are a few ways to pull yourself out of a bad mood before it gets too big to control.</p><h3>Pinpoint what’s gone wrong</h3><p>Rather than resign yourself to staying crabby, Hanley suggests looking for specific triggers, like getting a bad night’s sleep or having unresolved feelings about something someone said to you. “When you can look and see that there is a legitimate reason why you’re in a bad mood, it helps dispel that sense of doom and gloom,” she explains.</p><p>Triggers can be physical, too. For instance, I have a bad habit of drinking too much coffee and little else, turning me into a jittery, dehydrated mess. According to Hanley, a small detail like this can be a contributing factor to a bad mood—and one that might be easily overlooked.</p><p>Hanley suggests asking yourself when the bad mood started and what set it off. “Maybe you snapped at your co-worker because you have a headache that you’ve been trying to pretend isn’t really there,” she says. And when you look at why you have a headache — possibly caused by, say, ingesting five coffees — now you know something concrete you can do to help yourself feel better, such as drinking a glass of water, massaging your neck and temples, or taking a pain reliever. “It shifts you from reactive to proactive.”</p><h3>Analyze your emotions</h3><p>“Emotions are a secondary response,” says Lee J. Colan, founder of leadership consulting firm The L Group and author of <a href=\\"https://www.thelgroup.com/resources/power-positive-coaching\\"><em>The Power of Positive Coaching</em></a>. “There’s generally something underlying the raw emotion, and having awareness of that enables you to respond in the correct way.” When I yelled back at that other driver, I was reflexively reacting. But angrily lingering on the interaction for the rest of the day was a choice — and one that only aggravated my mood.</p><blockquote>“Even saying to yourself, ‘Wow, I’m in a bad mood,’ can be just the opening to ask yourself, ‘What do I most need right now?’”</blockquote><p>Give yourself the space to step back from your immediate emotional reactions and reflect on them, says Leslie Becker-Phelps, a licensed psychologist and author of <a href=\\"https://www.amazon.com/gp/product/B00KBEHMJ2/ref=dbs_a_def_rwt_hsch_vapi_tkin_p1_i0\\"><em>Insecure in Love</em></a>. By thinking critically about the situation and your role in it, you can create some distance from your emotions, which in turn allows you to gain more control over how you’re feeling.</p><h3>Practice finding the upside…</h3><p>When faced with a setback, even a small one — you forget your password for the umpteenth time, or a meeting gets rescheduled yet again — it’s easy to let yourself fall into a steady stream of complaints.</p><p>“The human brain is wired to look for threats,” Hanley says. Evolutionarily, this helped keep us alive — but in modern-day life, threats can look a lot more like minor annoyances, creating a <a href=\\"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3652533/\\">negativity bias</a> that keeps us focused on what’s going wrong. “To bust out of this neural rut,” she says, “train yourself to acknowledge when things go right.”</p><p>To that end, Colan recommends writing down three things that went well each day for two weeks and what your role was in each of those things. “It’s a way for you to say, ‘I had a positive impact,’” Colan says. “And I don’t care how much of a bummer your day is, there’s always a couple of things that went well.” The more attuned you are to this fact, the better equipped you’ll be to consciously respond with positivity the next time things go wrong.</p><h3>…And giving the benefit of the doubt</h3><p>When you feel powerless against the whims of the universe, it’s easy to overlook the one thing you do have control over: your empathy. “If we’re just reacting, we’re kind of taking our choice away,” Colan says. “We’re saying, ‘Here’s the stimulus. I’m reacting.’ But I think the more emotionally mature response is to say, ‘Alright, even though it feels quick, I have a choice here.’”</p><p>When you’re in a bad mood, choosing to empathize with the person who set you off can be a way of stopping yourself from stewing in your own hard feelings. But in the moment, being able to simply wave off an offense without reacting is much easier said than done.</p><p>“It’s really about creating a practice of empathy over time, which means identifying why the other person might feel the way they do,” Colan says. Think of it as a constant thought experiment. If a colleague’s giving you a hard time, for example, imagine what their own stressors might be — maybe they have a sick child at home, or a costly car repair, or something else you’d have no way of knowing about. The trick is to retrain your mind enough that this empathy becomes an immediate reflex, so that you can harness it before you let something derail your mood and your day. That’s a skill that comes with time and practice.</p><h3>Cut yourself some slack</h3><p>Personal change takes time and effort. Getting upset with yourself for not being able to change your emotional state will only make it worse. Instead, try using your bad mood as an excuse to practice a little self-compassion.</p><p>“You can’t change something — a habit, a mood, or a bad day — without acknowledging that you don’t like what’s currently happening,” Hanley says. “Even saying to yourself, ‘Wow, I’m in a bad mood,’ can be just the opening to ask yourself, ‘What do I most need right now?’”</p><p>Maybe you just need a break, and you can go outside for 10 minutes at lunch instead of running an errand or eating at your desk. Hanley recommends playing your favorite music on the way home instead of listening to the news.</p><p>But to make that happen, you have to admit that you need it in the first place. Make an effort to regularly check in with yourself throughout the day for what you need in a given moment — whether it’s a glass of water or a quick walk around the block. Colan says there’s a level of mindfulness to it: “One of the key things to get through any point of adversity—even if it’s just a bad mood or rough day—you have to plan for the future but live in the present.”</p><p>Taking that idea and zooming out with it, you can also use your past experience to help you: Look back at previous rough patches and see how everything ended up working out. You’ve gotten through plenty of bad moods in the past, and you’ll do it again — and reminding yourself that this is temporary may help speed the process along.</p></div></div></div></div></div>
        </body>
      </html>
    ",
      }
    }
    startInLoadingState={true}
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "flex": 0,
          "height": 0,
        },
      ]
    }
    useSharedProcessPool={true}
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F1F1",
          "bottom": 0,
          "flex": 1,
          "height": "100%",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
      ]
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
</View>
`;
